File: Task Tracker C#/contest/contestNetworking/ContestProtoWorker.cs
using Com.Example.Networking;
using contestService;
using Google.Protobuf;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Sockets;
using System.Runtime.Serialization;
using System.Runtime.Serialization.Formatters.Binary;
using System.Text;
using System.Threading.Tasks;
using Type = contestDomain.enums.Type;

namespace contestNetworking
{
    public class ContestProtoWorker : IContestObserver
    {

        private IContestService server;
        private TcpClient connection;

        private NetworkStream stream;
        private volatile bool connected;

        public ContestProtoWorker(IContestService server, TcpClient connection)
        {
            this.server = server;
            this.connection = connection;
            try
            {

                stream = connection.GetStream();
                connected = true;
            }
            catch (Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }
        }

        public virtual void run()
        {
            while (connected)
            {
                try
                {
                    ContestRequest request = ContestRequest.Parser.ParseDelimitedFrom(stream);
                    ContestResponse response = HandleRequest(request);
                    if (response != null)
                    {
                        SendResponse(response);
                    }
                }
                catch (Exception e)
                {
                    Console.WriteLine(e.StackTrace);
                }

                try
                {
                    Thread.Sleep(1000);
                }
                catch (Exception e)
                {
                    Console.WriteLine(e.StackTrace);
                }
            }
            try
            {
                stream.Close();
                connection.Close();
            }
            catch (Exception e)
            {
                Console.WriteLine("Error " + e);
            }
        }

        private ContestResponse HandleRequest(ContestRequest request)
        {
            ContestResponse response = null;
            ContestRequest.Types.Req requestType = request.Req;
            switch (requestType)
            {
                case ContestRequest.Types.Req.Login:
                    {
                        Console.WriteLine("Logging in...");
                        contestDomain.User user = ProtoUtils.GetUser(request);
                        Console.WriteLine("user " + user);
                        try
                        {
                            lock (server)
                            {
                                Console.WriteLine("before login");
                                server.Login(user, this);
                                Console.WriteLine("Logged in the suer");
                            }
                            return ProtoUtils.CreateOKResp();
                        }
                        catch (ContestException cx)
                        {
                            connected = false;
                            return ProtoUtils.CreateErrorResponse(cx.Message);
                        }
                    }
                case ContestRequest.Types.Req.Logout:
                    {
                        Console.WriteLine("Logging out...");
                        contestDomain.User user = ProtoUtils.GetUser(request);
                        try
                        {
                            lock (server)
                            {
                                server.Logout(user, this);
                            }
                            connected = false;
                            return ProtoUtils.CreateOKResp();
                        }
                        catch (ContestException cx)
                        {
                            return ProtoUtils.CreateErrorResponse(cx.Message);
                        }
                    }
                case ContestRequest.Types.Req.AddParticipantExperimental:
                    {
                        Console.WriteLine("Adding participant...");
                        contestDomain.ParticipantTasks participant = ProtoUtils.GetPT(request);
                        try
                        {
                            lock (server)
                            {
                                server.Register(participant.Name, participant.Age, participant.Type1, participant.Type2);
                            }
                            return ProtoUtils.CreateRegisterResponse();
                        }
                        catch (ContestException cx)
                        {
                            return ProtoUtils.CreateErrorResponse(cx.Message);
                        }
                    }
                case ContestRequest.Types.Req.GetAllTasksExperimental:
                    {
                        Console.WriteLine("Getting all tasks...");
                        List<contestDomain.dto.TaskDTO> tk = new List<contestDomain.dto.TaskDTO>();
                        try
                        {
                            lock (server)
                            {
                                tk = server.GetTasksEnrolled();
                            }
                            return ProtoUtils.CreateGetTasksEnrolledResponse(ProtoUtils.GetTasksConverted(tk));
                        }
                        catch (ContestException cx)
                        {
                            return ProtoUtils.CreateErrorResponse(cx.Message);
                        }
                    }
                case ContestRequest.Types.Req.Filter:
                    {
                        Console.WriteLine("Filtering participants...");
                        var filters = ProtoUtils.GetFilters(request);
                        List<contestDomain.Participant> filteredParticipants = new List<contestDomain.Participant>();
                        try
                        {
                            lock (server)
                            {
                                filteredParticipants = server.GetFilteredParticipants((Type)Enum.Parse(typeof(Type), filters[0]), (contestDomain.enums.AgeGroup)Enum.Parse(typeof(contestDomain.enums.AgeGroup), filters[1]));
                            }
                            return ProtoUtils.CreateGetFilteredParticipantsResponse(ProtoUtils.GetFilteredParticipantsDTO(filteredParticipants));
                        }
                        catch (ContestException cx)
                        {
                            return ProtoUtils.CreateErrorResponse(cx.Message);
                        }
                    }
            }
            return response;
        }

        private void SendResponse(ContestResponse response)
        {
            Console.WriteLine("Sending response " + response);
            lock (stream)
            {
                response.WriteDelimitedTo(stream);
                stream.Flush();
            }
            Console.WriteLine("Sent");
        }

        public void updateTaskList()
        {
            Console.WriteLine("Updating tables...");
            try
            {
                SendResponse(ProtoUtils.CreateUpdateResponse());
            }
            catch (ContestException cx)
            {
                Console.WriteLine(cx.Message);
            }
        }
    }
}

File: Task Tracker Java/contest_java/contestNetworking/src/main/java/com/example/protobuff/ContestProtoWorker.java
package com.example.protobuff;

import com.example.ContestException;
import com.example.IContestObserver;
import com.example.IContestService;
import com.example.domain.User;

import java.io.*;
import java.net.Socket;

public class ContestProtoWorker implements Runnable, IContestObserver {
    private IContestService server;
    private Socket connection;
    private InputStream input;
    private OutputStream output;
    private volatile boolean connected;

    public ContestProtoWorker(IContestService server, Socket connection) {
        this.server = server;
        this.connection = connection;
        try {
            output = connection.getOutputStream();
            output.flush();
            input = connection.getInputStream();
            connected = true;
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void updateTaskList() throws ContestException {
        System.out.println("Updating tasks list...");
        try {
            sendResponse(ProtoUtils.createUpdateResponse());
        } catch (Exception e) {
            throw new RuntimeException(e);
        }

    }

    @Override
    public void run() {
        while (connected) {
            try {
                System.out.println("Awaiting requests...");
                ContestProto.ContestRequest request = ContestProto.ContestRequest.parseDelimitedFrom(input);
                System.out.println("Request received " + request);
                ContestProto.ContestResponse response = handleRequest(request);
                if (response != null) {
                    sendResponse(response);
                }
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
            try {
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
        try {
            input.close();
            output.close();
            connection.close();
        } catch (IOException e) {
            System.out.println("Error " + e);
        }
    }

    private void sendResponse(ContestProto.ContestResponse response) throws IOException {
        System.out.println("Sending response " + response);
        synchronized (output) {
            response.writeDelimitedTo(output);
            output.flush();
        }
    }

    private ContestProto.ContestResponse handleRequest(ContestProto.ContestRequest request) {
        ContestProto.ContestResponse response = null;
        switch (request.getReq()) {
            case Login:
                System.out.println("Login request...");
                User user = ProtoUtils.getUser(request);
                try {
                    server.login(user, this);
                    return ProtoUtils.createOKResponse();
                } catch (IOException | ContestException e) {
                    connected = false;
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }
            case Logout:
                System.out.println("Logout request...");
                User user1 = ProtoUtils.getUser(request);
                try {
                    server.logout(user1, this);
                    return ProtoUtils.createOKResponse();
                } catch (ContestException e) {
                    connected = false;
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }
            case ADD_PARTICIPANT_EXPERIMENTAL:
                System.out.println("Registering participant request...");
                var pt = ProtoUtils.getPT(request);
                try {
                    server.register(pt.getName(), pt.getAge(), pt.getType1(), pt.getType2());
                    return ProtoUtils.createRegisterResponse();
                } catch (IOException | ContestException e) {
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }
            case GET_ALL_TASKS_EXPERIMENTAL:
                System.out.println("Getting all the tasks request...");
                try {
                    return ProtoUtils.createGetTasksEnrolledResponse(ProtoUtils.getEnrolledTasksForResponse(server.getTasksEnrolled()));
                } catch (ContestException e) {
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }
            case FILTER:
                System.out.println("Filtering participants request...");
                var filters = ProtoUtils.getFilters(request);
                try {
                    var pdt = ProtoUtils.getFilteredPartiticipants(server.getFilteredParticipants(com.example.enums.Type.valueOf(filters[0]), com.example.enums.AgeGroup.valueOf(filters[1])));
                    return ProtoUtils.createGetFilteredParticipantsResponse(ProtoUtils.getFilteredPartsDTOMapper(pdt));
                } catch (ContestException | IllegalArgumentException e) {
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }
        }
        return response;
    }
}
